Add optional support for HttpClient

bedag-moo opened the pull request:
Support HttpClient in addition to Http, clean up generated code

Fixes #6080 

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: master for non-breaking changes and `3.0.0` branch for breaking (non-backward compatible) changes.

wing328 commented:
cc @Vrolijkx @taxpon @sebastianhaas @kenisteward @TiFu

bedag-moo commented:
@wing328: Can you clarify what up with these branches? I thought since most typescript commits were only in 2.3.0, I should base my PR on that branch to minimize merging pain when 2.3.0 is finally merged?

sebastianhaas commented:
As the PR checklist says, you should
> File[d] the PR against the correct branch: master for non-breaking changes and 3.0.0 branch for breaking (non-backward compatible) changes.

wing328 commented:
@bedag-moo the latest master of 2.3.0 (snapshot) so that's why I updated the PR to file against master.

bedag-moo commented:
Ah, 2.3.0 has already been merged! That would have been nice to know before starting work on this PR (Because 2.3.0 is not released yet, I had assumed that master would remain for 2.2.x releases, as has been the case for nearly a year now). 

Might I suggest describing the mapping between branches and releases on your contribution page? And to delete branches after merging them?

Meanwhile, I'll rebase my changes to master.

kenisteward commented:
Agreed. I thought 2.3.0 wasn't merged yet as well

On Tue, Sep 12, 2017, 1:19 PM Adrian Moos <notifications@github.com> wrote:

> Ah, 2.3.0 has already been merged! That would have been nice to know
> before starting work on this PR (Because 2.3.0 is not released yet, I had
> assumed that master would remain for 2.2.x releases, as has been the case
> for nearly a year now).
>
> Might I suggest describing the mapping between branches and releases on
> your contribution page? And to delete branches after merging them?
>
> Meanwhile, I'll rebase my changes to master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swagger-api/swagger-codegen/pull/6295#issuecomment-328922542>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMPLtdHl9w07p7I39IwP2tkp2AQ51-ICks5shr0mgaJpZM4O1Hzu>
> .
>

bedag-moo commented:
Rebase complete  (I had to do it manually, because everything had been moved to a different directory in the meantime).

topce commented:
@bedag-moo Nice work.
One small issue for response type blob it does not compile 
`return this.httpClient.get<any>(`${this.basePath}/content/bulk`, {
            params: queryParameters,
            headers: headers,
            responseType: "blob",
            withCredentials: this.configuration.withCredentials,
        });`

tsc errror is 

 error TS2345: Argument of type '{ params: HttpParams; headers: HttpHeaders; responseType: "blob"; withCredentials: bool
ean | unde...' is not assignable to parameter of type '{ headers?: HttpHeaders | undefined; observe?: "body" | undefined; params?: HttpParams | undefine...'.
  Type '{ params: HttpParams; headers: HttpHeaders; responseType: "blob"; withCredentials: boolean | unde...' is not assignable to type '{ headers?: HttpHeaders | undefined;
observe?: "body" | undefined; params?: HttpParams | undefine...'.
    Types of property 'responseType' are incompatible.
      Type '"blob"' is not assignable to type '"json" | undefined'.
but code bellow works (removed <any>)
`return this.httpClient.get(`${this.basePath}/content/bulk`, {
            params: queryParameters,
            headers: headers,
            responseType: "blob",
            withCredentials: this.configuration.withCredentials,
        });`

bedag-moo commented:
Thanks for the bug report; should be fixed now.

wing328 commented:
> Might I suggest describing the mapping between branches and releases on your contribution page? And to delete branches after merging them?

Definitely. Let me try this weekend.

sebastianhaas commented on modules/swagger-codegen/src/main/java/io/swagger/codegen/utils/SemVer.java:
```diff
@@ -0,0 +1,33 @@
+package io.swagger.codegen.utils;
```
I'd also vote for that :)

bedag-moo replied:
Done. It is the first unit test in this package.

kenisteward commented:
Did we make sure that this only applies in angular 4.3 and update the peer
dependencies as such ?

On Thu, Sep 14, 2017, 5:00 AM Sebastian Haas <notifications@github.com>
wrote:

> *@sebastianhaas* commented on this pull request.
> ------------------------------
>
> In
> modules/swagger-codegen/src/main/java/io/swagger/codegen/utils/SemVer.java
> <https://github.com/swagger-api/swagger-codegen/pull/6295#discussion_r138835172>
> :
>
> > @@ -0,0 +1,33 @@
> +package io.swagger.codegen.utils;
>
> I'd also vote for that :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swagger-api/swagger-codegen/pull/6295#discussion_r138835172>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMPLtamJaquGdc_VMoR7SnIyuPAs6Bbjks5siOsugaJpZM4O1Hzu>
> .
>

sebastianhaas commented:
@kenisteward check out [this line](https://github.com/swagger-api/swagger-codegen/pull/6295/files#diff-ad60e100e050d5ef52943803a2adc5fdR115).

kenisteward commented:
I missed the Ng version check. :) I'm very happy with this change

On Thu, Sep 14, 2017, 8:30 AM Sebastian Haas <notifications@github.com>
wrote:

> @kenisteward <https://github.com/kenisteward> check out this line
> <https://github.com/swagger-api/swagger-codegen/pull/6295/files#diff-ad60e100e050d5ef52943803a2adc5fdR115>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swagger-api/swagger-codegen/pull/6295#issuecomment-329466938>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMPLtUV6LohGTCpobPIZBDqGtKaCa-BHks5siRx4gaJpZM4O1Hzu>
> .
>

wing328 commented:
> Definitely. Let me try this weekend.
 
UPDATE: renamed 2.3.0 to 2.3.0_BACKUP_ONLY (to keep the git log for changes made to 2.3.0)

wing328 commented:
Restarted Travis CI job. Will merge after the test passes.

macjohnny commented:
@bedag-moo it seems there are some conflicts with the changes introduced in https://github.com/swagger-api/swagger-codegen/pull/6525

I suggest you try to adapt your PR to conform with the fixes introduced in https://github.com/swagger-api/swagger-codegen/pull/6525, as replacing path-parameters with their value without url-escaping can lead to problems, see https://github.com/swagger-api/swagger-codegen/issues/6524

krizic commented:
Generated functions are always having a trailing comma at the end! Please provide the fix.

```
getBillingInformation(start: Date, end: Date,): Observable<Array<BillingInformation>>{}
downloadLogoForCompany(companyId: string, fetchThumb: boolean,): Observable<ArrayBuffer>{}
...
```

sebastianhaas commented:
I‘ll try to provide Angular Integration Tests this week so we can avoid these kind of issues.

> On 27.09.2017, at 17:08, Vedran Krizic <notifications@github.com> wrote:
> 
> Generated functions are always having a trailing comma at the end! Please provide the fix.
> 
> getBillingInformation(start: Date, end: Date,): Observable<Array<BillingInformation>>{}
> downloadLogoForCompany(companyId: string, fetchThumb: boolean,): Observable<ArrayBuffer>{}
> ...
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

wing328 commented:
@bedag-moo I wonder if you can rebase on the latest master to resolve the merge conflicts (sorry I should have merged yours first)

kenisteward commented:
@krizic is this due to these changes or outside of these changes?  If it is outside these changes probably better to make a different issue.

krizic commented on modules/swagger-codegen/src/main/resources/typescript-angular/api.service.mustache:
```diff
@@ -86,12 +79,8 @@ export class {{classname}} {
-    public {{nickname}}({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}, {{/allParams}}extraHttpRequestParams?: any): Observable<{{#returnType}}{{{returnType}}}|undefined{{/returnType}}{{^returnType}}{}{{/returnType}}> {
-{{/isResponseFile}}
+    {{! if you change this method signature, also change the version below }}
+    public {{nickname}}({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}, {{/allParams}}{{^useHttpClient}}extraHttpRequestParams?: RequestOptionsArgs{{/useHttpClient}}): Observable<{{#returnType}}{{{returnType}}}{{#isResponseTypeFile}}|undefined{{/isResponseTypeFile}}{{/returnType}}{{^returnType}}{}{{/returnType}}> {
```
This causes to always have a trailing comma in the end of all of the parameters.

krizic replied:
As mentioned above it produces this result:
```
getBillingInformation(start: Date, end: Date,): Observable<Array<BillingInformation>>{}
downloadLogoForCompany(companyId: string, fetchThumb: boolean,): Observable<ArrayBuffer>{}
```

kenisteward replied:
Yup that definitely needs to be fixed then.

wing328 replied:
@krizic thanks for testing the change.

I think the following should fix the issue
```
    public {{nickname}}({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}{{#hasMore}}, {{/hasMore}}{{/allParams}}{{^useHttpClient}}, extraHttpRequestParams?: RequestOptionsArgs{{/useHttpClient}}): Observable<{{#returnType}}{{{returnType}}}{{#isResponseTypeFile}}|undefined{{/isResponseTypeFile}}{{/returnType}}{{^returnType}}{}{{/returnType}}> {
 ```

When you've time, I wonder if you can give it a try by replacing line 83 with it.

cc @bedag-moo

krizic replied:
I can confirm this is still an issue, and if you see the code it is obvious that in case that `useHtpClient` flag is true he WON'T add last method parameter (extraHttpRequestParams) effectively leaving method signature to always end with `,`.

`{{^useHttpClient}}*extraHttpRequestParams?:RequestOptionsArgs{{/useHttpClient}}`

krizic replied:
@wing328 thanks for providing a quick solution!

wing328 replied:
I added `{{#hasMore}}, {{/hasMore}}` to skip "," for the last parameter. Are you using `-t` parameter to supply the customized templates with the fix?

wing328 replied:
Otherwise, you will need to do a "mvn clean package" to recompile the JAR.

krizic replied:
@wing328 you are right. I wasn't cleaning bins. Works now, thanks for the solution again. I hope this gets merged soon

wing328 replied:
@krizic np. Thanks for confirming the fix works.

krizic commented:
`CRITICAL` : PATCH request is missing body parameter so it sends the whole httpParams as the body! 

```
    return this.httpClient.patch<any>(`${this.basePath}v1/risk`, {
      params:          queryParameters,
      headers:         headers,
      withCredentials: this.configuration.withCredentials,
    });
```
actual signature of httpClient.Patch:
```
    patch(url: string, body: any | null, options?: {}): Observable<Object>;
```

bedag-moo commented:
Yes, I really would have preferred you to merge my changes first:

- I can not merge or rebase current master, because 3ac2b803f972d3919f1b8d04e0b32ef20613b15d removed and added a file differing only in file name case in a single commit, making it impossible to apply this commit on a case-insensitive filesystem (windows). Fortunately, there were no conflicts after this point.
- There is a merge conflict with #6454, whose intent is poorly documented (why would I set an accept header to what the server supports? The purpose of the accept header is telling the server what the *client* accepts!), and implemented very strangely (Why set the content type header to JSON, but only when posting a form? forms are never JSON ...). Since I see no value in this code, I have removed it. If you see a value, feel free to add it back in after fixing its bugs.
- There was a merge conflict with #6524, but at least here the intent was clear and the implementation sound, so adapting it was easy.

Please merge soon ...

bedag-moo commented:
Oh no, there was yet another conflict behind the commit that can not be applied on windows ...

@wing328 : Can you fix that one?

macjohnny commented:
@bedag-moo 
> There is a merge conflict with #6454, whose intent is poorly documented (why would I set an accept header to what the server supports? The purpose of the accept header is telling the server what the client accepts!), and implemented very strangely (Why set the content type header to JSON, but only when posting a form? forms are never JSON ...). Since I see no value in this code, I have removed it. If you see a value, feel free to add it back in after fixing its bugs.

I totally agree with you. The PR #6454 also resulted in a new bug: https://github.com/swagger-api/swagger-codegen/issues/6588, which would be fixed by merging your PR #6295

@wing328 please merge this PR as soon as possible

wing328 commented:
> @wing328 : Can you fix that one? You should be able to push to my branch.

Fixed but now encounter another issue due to methods with no argument, which result in the following errors:
```
api/store.service.ts(87,25): error TS1138: Parameter declaration expected.
api/store.service.ts(160,37): error TS1138: Parameter declaration expected.
api/user.service.ts(168,23): error TS1138: Parameter declaration expected.
api/user.service.ts(373,35): error TS1138: Parameter declaration expected.
```

I'm trying to come up with a fix.

macjohnny commented:
@wing328 isn't this related to https://github.com/swagger-api/swagger-codegen/pull/6295#pullrequestreview-65826900 ?

wing328 commented:
@bedag-moo yup and I got a fix. Testing it now

wing328 commented:
After fixing that issue, now I got the following errors when running `npm run build`:
```
api/pet.service.ts(272,77): error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
api/pet.service.ts(380,77): error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
api/pet.service.ts(468,77): error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
api/pet.service.ts(521,77): error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
```

wing328 commented:
Fixed wth `String()` and the CI tests passed: https://travis-ci.org/swagger-api/swagger-codegen/builds/281593760

wing328 commented:
82b6e81 has the fixes for the issues mentioned above.

Please review when you've time.

cc @TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx

kenisteward commented:
@wing328 I looked over the string() addition and it looks safe. If that's all you needed I approve .  On my phone so I cant tell how to approve a change haha

wing328 commented:
@kenisteward thanks for reviewing the change. Have a nice weekend!

xmeng1 commented:
+1

kenisteward commented:
This is already possible with 2.3.0

On Thu, Jan 18, 2018, 6:29 AM Xin Meng <notifications@github.com> wrote:

> +1
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swagger-api/swagger-codegen/pull/6295#issuecomment-358619643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMPLtU2oHAI9jpc5CHsMjfx51LaNb4XYks5tLysmgaJpZM4O1Hzu>
> .
>

takahashi64 commented:
Is it possible to get swagger codegen cli to set useHttpClient to false?

kenisteward commented:
It is set to false by default

On Mon, Jan 29, 2018, 8:20 AM John Lucey <notifications@github.com> wrote:

> Is it possible to get swagger codegen cli to set useHttpClient to false?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swagger-api/swagger-codegen/pull/6295#issuecomment-361242844>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMPLtSHy8BkQ10QPuHaqKuYaw28m1yBjks5tPcWBgaJpZM4O1Hzu>
> .
>

kenisteward commented:
It's set based on version you provide actually sorry. So just define with
any version under 4.3

On Mon, Jan 29, 2018, 9:01 AM Kenneth Steward <keni.steward@gmail.com>
wrote:

> It is set to false by default
>
> On Mon, Jan 29, 2018, 8:20 AM John Lucey <notifications@github.com> wrote:
>
>> Is it possible to get swagger codegen cli to set useHttpClient to false?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/swagger-api/swagger-codegen/pull/6295#issuecomment-361242844>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AMPLtSHy8BkQ10QPuHaqKuYaw28m1yBjks5tPcWBgaJpZM4O1Hzu>
>> .
>>
>
